plugins/woocommerce/src/Internal/Utilities/Types.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Automattic\WooCommerce\Internal\Utilities;
+
+use InvalidArgumentException;
+
+/**
+ * Utilities to help ensure type safety.
+ */
+class Types {
+	public const UNABLE_TO_ENSURE_INSTANCE_IS_OF_EXPECTED_TYPE = 'unable-to-ensure-instance-is-of-expected-type';
+
+	/**
+	 * Checks if $thing is an instance of $desired_type.
+	 *
+	 * If the check succeeds, $thing will be returned without further modification. If the check fails, then either
+	 * an exception will be thrown or, if an $on_failure callback was supplied, it will be invoked to either generate
+	 * an appropriate return value or to throw a more specific exception.
+	 *
+	 * Please note that the failure handler will be passed two arguments:
+	 *
+	 *     $on_failure( $object, $desired_type )
+	 *
+	 * @since 9.1.0
+	 * @throws InvalidArgumentException If $object does not match $desired_type, and an $on_failure callback was not supplied.
+	 *
+	 * @param mixed     $thing        The value or reference to be assessed.
+	 * @param string    $desired_type What we expect the return type to be, if it is not a WP_Error.
+	 * @param ?callable $on_failure   If provided, and if evaluation fails, this will be invoked to generate a return value.
+	 *
+	 * @return mixed
+	 */
+	public static function ensure_instance_of( $thing, string $desired_type, callable $on_failure = null ) {
```

Comment:
Another thought is, `$thing` could optionally be an array (to support the equivalent of a type union, like `WC_Order|WC_Super_Order`) ... but, that could also be a thing to keep in reserve until it's needed (we have additional latitude to change how this works in future releases, since it is in the internal namespace).

Reply:
🤔 Do you mean `$desired_type` could be an array?

Reply:
Sorry, yes, that's what I meant.

Reply:
I will not make this change (yet), because:

- [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).
- We can add it later (we are within the internal namespace, so we have that degree of latitude).